src/Processors/QueryPlan/QueryPlan.cpp

```diff
@@ -455,16 +455,24 @@ static void updateDataStreams(QueryPlan::Node & root)
 
         static bool visitTopDownImpl(QueryPlan::Node * /*current_node*/, QueryPlan::Node * /*parent_node*/) { return true; }
 
-        static void visitBottomUpImpl(QueryPlan::Node * current_node, QueryPlan::Node * parent_node)
+        static void visitBottomUpImpl(QueryPlan::Node * current_node, QueryPlan::Node * /*parent_node*/)
         {
-            if (!parent_node || parent_node->children.size() != 1)
+            auto & current_step = *current_node->step;
+            if (!current_step.canUpdateInputStream() || current_node->children.empty())
                 return;
 
-            if (!current_node->step->hasOutputStream())
-                return;
+            for (const auto * child : current_node->children)
+            {
+                if (!child->step->hasOutputStream())
+                    return;
+            }
+
+            DataStreams streams;
+            streams.reserve(current_node->children.size());
+            for (const auto * child : current_node->children)
+                streams.emplace_back(child->step->getOutputStream());
 
-            if (auto * parent_transform_step = dynamic_cast<ITransformingStep *>(parent_node->step.get()); parent_transform_step)
-                parent_transform_step->updateInputStream(current_node->step->getOutputStream());
+            current_step.updateInputStreams(std::move(streams));
```

Comment:
I found that the problem was because `updateDataStreams()` didn't update the input and output streams of `UnionStep` after optimizations have been applied to the query plan. It used to update only query plan's steps derived from `ITransformingStep` and `UnionStep` is not derived from `ITransformingStep` (because `UnionStep` has more than 1 argument).